File: lox/token.cpp
#include <lox/token.hpp>

#include <fmt/std.h>

#include <utility>

namespace lox {

constexpr std::array token_type_names{
    "LEFT_PAREN", "RIGHT_PAREN", "LEFT_BRACE", "RIGHT_BRACE", "COMMA", "DOT",
    "MINUS", "PLUS", "SEMICOLON", "SLASH", "STAR",

    // One or two character tokens
    "BANG", "BANG_EQUAL", "EQUAL", "EQUAL_EQUAL", "GREATER", "GREATER_EQUAL",
    "LESS", "LESS_EQUAL",

    // Literals
    "IDENTIFIER", "STRING", "NUMBER",

    // Keywords
    "AND", "CLASS", "ELSE", "FALSE", "FUN", "FOR", "IF", "NIL", "OR", "PRINT",
    "RETURN", "SUPER", "THIS", "TRUE", "VAR", "WHILE",

    "EOF"};

// Ensure that all token types can be looked up
static_assert(size(token_type_names) ==
              static_cast<std::size_t>(token_type::NUM_TYPES));

constexpr auto to_string(token_type type) -> const char* {
  return token_type_names.at(static_cast<std::size_t>(type));
}

// clang-format off
template <class... Ts> struct overloaded : Ts... { using Ts::operator()...; }; 
template <class... Ts> overloaded(Ts...) -> overloaded<Ts...>;
// clang-format on

auto to_string(token_literal literal) -> std::string {
  using namespace std::string_literals;
  return std::visit(
      overloaded{[](std::monostate) { return "null"s; },
                 [](bool arg) { return arg ? "true"s : "false"s; },
                 [](double arg) { return fmt::format("{}", arg); },
                 [](std::string arg) { return fmt::format("\"{}\"", arg); }},
      literal);
}

auto token::str() const -> std::string {
  return literal.index() == 0U
             ? fmt::format("{}", to_string(type))
             : fmt::format("{} {} {}", to_string(type), lexeme, literal);
}

} // namespace lox

File: glox/internal/token/token.go
package token

import (
	"fmt"
)

//go:generate stringer -type=Type -linecomment
type Type int

const (
	// Single-character tokens
	LeftParen Type = iota
	RightParen
	LeftBrace
	RightBrace
	Comma
	Dot
	Minus
	Plus
	Semicolon
	Slash
	Star
	QuestionMark
	Colon

	// One or two character tokens
	Bang
	BangEqual
	Equal
	EqualEqual
	Greater
	GreaterEqual
	Less
	LessEqual

	// Literals
	Identifier
	String
	Number

	// Keywords
	keywordStart
	And    // and
	Class  // class
	Else   // else
	False  // false
	Fun    // fun
	For    // for
	If     // if
	Nil    // nil
	Or     // or
	Print  // print
	Return // return
	Super  // super
	This   // this
	True   // true
	Var    // var
	While  // while
	keywordEnd

	EOF
	numTokens
)

var keywords = map[string]Type{}

func init() {
	// Initialise keywords
	for i := keywordStart + 1; i < keywordEnd; i++ {
		keywords[i.String()] = i
	}
}

func LookupKeyword(identifier string) Type {
	if ttype, ok := keywords[identifier]; ok {
		return ttype
	}
	return Identifier
}

type Token struct {
	Type   Type
	Lexeme string
	Line   int
}

// Debug converts the token into a string useful for debugging.
func (t Token) Debug() string {
	switch ty := t.Type; ty {
	case EOF:
		return "EOF"
	case String:
		return fmt.Sprintf("%s(%q)", ty, t.Lexeme)
	case Identifier, Number:
		return fmt.Sprintf("%s(%v)", ty, t.Lexeme)
	default:
		return ty.String()
	}
}

// String converts the token to a string useful for the user.
func (t Token) String() string {
	switch t.Type {
	case EOF:
		return "EOF"
	default:
		return t.Lexeme
	}
}
